aten/src/ATen/native/kleidiai/kai_kernels.cpp

```diff
@@ -0,0 +1,442 @@
+    for (int64_t x = begin; x < end; x++) {
+      matmul_groupwise(
+          std::ref(ukernel),
+          m,
+          num_n_per_thread,
+          x * num_n_per_thread,
+          k,
+          bl,
+          dst_stride,
+          lhs_packed,
+          rhs_packed_mtx_qs4cx,
+          dst_act_mtx_f32);
+    }
+  });
+
+  delete[] lhs_packed;
+}
+
+static void kai_quant_pack_lhs_int4_mm_channelwise(
+    const Tensor& output,
+    const Tensor& input,
+    const Tensor& weight,
+    const int64_t m,
+    const int64_t n,
+    const int64_t k) {
+  // Kernel IDs for GEMM and GEMV
+  kai_kernel_id gemm_id =
+      kai_kernel_id::matmul_clamp_f32_qai8dxp4x8_qsi4cxp8x8_8x8x32_neon_i8mm;
+  kai_kernel_id gemv_id =
+      kai_kernel_id::matmul_clamp_f32_qai8dxp1x8_qsi4cxp8x8_1x8x32_neon_dotprod;
+
+  // Get the total number of threads available and choose GEMM or GEMV steps
+  const int64_t total_threads = at::get_num_threads();
+  auto gemm_kernel_packet = kai_select_channelwise_matmul_ukernel(gemv_id);
+  if (cpuinfo_has_arm_i8mm()) {
+    gemm_kernel_packet = kai_select_channelwise_matmul_ukernel(gemm_id);
+  }
+  auto gemv_kernel_packet = kai_select_channelwise_matmul_ukernel(gemv_id);
+
+  // Retrieve m_step and n_step values from GEMM and GEMV kernels
+  const int64_t gemm_m_step = gemm_kernel_packet.ukernel.get_m_step();
+  const int64_t gemm_n_step = gemm_kernel_packet.ukernel.get_n_step();
+  const int64_t gemv_m_step = gemv_kernel_packet.ukernel.get_m_step();
+  const int64_t gemv_n_step = gemv_kernel_packet.ukernel.get_n_step();
+  // Determine threading and kernel type
+  ThreadDivision division = get_thread_division(
+      total_threads,
+      m,
+      n,
+      k,
+      gemm_m_step,
+      gemm_n_step,
+      gemv_m_step,
+      gemv_n_step);
+  // Select appropriate kernel packet based on the chosen kernel type
+  auto& kernel_packet =
+      division.use_gemm ? gemm_kernel_packet : gemv_kernel_packet;
+
+  // Thread blocking parameters
+  const size_t mr = kernel_packet.ukernel.get_mr();
+  const size_t nr = kernel_packet.ukernel.get_nr();
+  const size_t kr = kernel_packet.ukernel.get_kr();
+  const size_t sr = kernel_packet.ukernel.get_sr();
+  const size_t m_increment = kernel_packet.ukernel.get_m_step();
+  const size_t n_per_thread = n / division.num_threads_x;
+  const size_t m_per_thread = m / division.num_threads_y;
+  const int64_t num_threads = division.num_threads_y * division.num_threads_x;
+  const size_t dst_stride = n * sizeof(float);
+  const size_t lhs_stride = k * sizeof(float);
+
+  const size_t lhs_packed_size =
+      kernel_packet.kai_get_lhs_packed_size(m_increment, k, mr, kr, sr);
+
+  uint8_t* dst_act_mtx_f32 = reinterpret_cast<uint8_t*>(output.data_ptr());
+  uint8_t* lhs_native_mtx_f32 = reinterpret_cast<uint8_t*>(input.data_ptr());
+  uint8_t* rhs_packed_mtx_qs4cx = reinterpret_cast<uint8_t*>(weight.data_ptr());
+  uint8_t* lhs_packed = new uint8_t[lhs_packed_size * num_threads];
```

Comment:
Please use managed pointers
```suggestion
  auto lhs_packed = std::make_unique<uint8_t[]>(lhs_packed_size * num_threads);
```

Reply:
done